### ISSUE TITLE
CognitiveComplexMethod doesn't take into account `when` cases

### ISSUE THREAD

**Issue by haluzpav** [detekt/detekt#5565]
## Expected Behavior
Each case of `when` statetment should increase "cognitive complexity" by 1.

## Observed Behavior
High number of `when` cases seem to have no effect on CognitiveComplexMethod. However the old CyclomaticComplexMethod does get triggered.

## Steps to Reproduce

```yml
complexity:
  CognitiveComplexMethod:
    active: true
```

```kotlin
fun giantWhen(): Int = when (Random.nextInt()) {
    0 -> 1
    1 -> 2
    2 -> 3
    3 -> 4
    4 -> 5
    5 -> 6
    6 -> 7
    7 -> 8
    8 -> 9
    9 -> 10
    10 -> 11
    11 -> 12
    12 -> 13
    13 -> 14
    14 -> 15
    15 -> 16
    16 -> 17
    17 -> 18
    18 -> 19
    19 -> 20
    20 -> 21
    21 -> 22
    22 -> 23
    23 -> 24
    24 -> 25
    25 -> 26
    26 -> 27
    27 -> 28
    28 -> 29
    29 -> 30
    30 -> 31
    31 -> 32
    32 -> 33
    33 -> 34
    34 -> 35
    35 -> 36
    36 -> 37
    37 -> 38
    38 -> 39
    39 -> 40
    40 -> 41
    else -> error("🤷‍")
}
```

Detekt output:

```
<path>: The function giantWhen appears to be too complex based on Cyclomatic Complexity (complexity: 43). Defined complexity threshold for methods is set to '30' [CyclomaticComplexMethod]
```

Where `CognitiveComplexMethod`?

(`CognitiveComplexMethod` does get triggered in some other cases which don't depend on `when`.)

## Your Environment

* Version of detekt used: 1.22
* Version of Gradle used (if applicable): 7.5.1
* Gradle scan link (add `--scan` option when running the gradle task): -
* Operating System and version: -
* Link to your project (if it's a public repository): -

cc @arturbosch @sanggggg

**Comment by sanggggg** [comment]
In paper of [CognitiveComplexity](https://www.sonarsource.com/docs/CognitiveComplexity.pdf),
> Under Cyclomatic Complexity, a switch is treated as an analog to an if-else if chain.
That is, each case in the switch causes an increment because it causes a branch in the
mathematical model of the control flow.
But from a maintainer’s point of view, a switch - which compares a single variable to an
explicitly named set of literal values - is much easier to understand than an if-else if
chain because the latter may make any number of comparisons, using any number of
variables and values.
In short, an if-else if chain must be read carefully, while a switch can often be taken
in at a glance

example below.
![image](https://user-images.githubusercontent.com/37951125/203689529-92bbddcf-8e01-48ca-8a06-de9ac69e7914.png)


So it is the intended behavior that all cases of `when` (~= `switch`) are not counted according to the specifications of the above paper.

**Comment by sanggggg** [comment]
However, sometimes `when` is actually recognized in a similar form to `if else` like below,
I think we can modify it more appropriately

```kotlin
// when used as a replacement for if else
when {
    x.isOdd() -> print("x is odd")
    y.isEven() -> print("y is even")
    else -> print("x+y is odd")
}
```

**Comment by haluzpav** [comment]
> So it is the intended behavior that all cases of when (~= switch) are not counted according to the specifications of the above paper.

Hm, alright. But I still think each case still does add some complexity, and it often has to be considered when working with such code.

As a side note, is it planned that `CognitiveComplexMethod` rule will eventually replace `CyclomaticComplexMethod`?

**Comment by BraisGabin** [comment]
> As a side note, is it planned that `CognitiveComplexMethod` rule will eventually replace `CyclomaticComplexMethod`?

No, I don't think so. They are two different ways to measure your function complexity.

---

About the real issue I don't have a strong opinion here. In your snipped I don't feel that the `when` add any cognitive complexity. It is just a `map`. It is cyclomatic complex because to check every single case I need a lot of tests but it's easy to read and understand what it does. I also agree that this could be an example and in the real life it could be a bit more difficult.

I would stick to the papers and what they say.
